junit-jupiter-api/src/main/java/org/junit/jupiter/api/parallel/ResourceLocksProvider.java

```diff
@@ -0,0 +1,187 @@
+ * are used then shared resources from them are summed up.
+ * It means that if resource 'A' is declared via {@code @ResourceLock}
+ * and resource 'B' added via {@code ResourceLocksProvider}
+ * then a test class or test method will have both resources: 'A' and 'B'.
+ *
+ * @since 5.12
+ * @see Isolated
+ * @see Resources
+ * @see ResourceAccessMode
+ * @see ResourceLock
+ * @see ResourceLocksFrom
+ * @see Lock
+ */
+@API(status = EXPERIMENTAL, since = "5.12")
+public interface ResourceLocksProvider {
+
+	/**
+	 * Add shared resources to a test class.
+	 *
+	 * @param testClass a test class to add shared resources
+	 * @return a set of {@link Lock}; may be empty
+	 *
+	 * @apiNote has the same semantics as adding {@code @ResourceLock} to a test class.
+	 */
+	default Set<Lock> provideForClass(Class<?> testClass) {
+		return emptySet();
+	}
+
+	/**
+	 * Add shared resources to a nested test class.
+	 *
+	 * @param testClass a nested test class to add shared resources
+	 * @return a set of {@link Lock}; may be empty
+	 *
+	 * @apiNote has the same semantics as adding {@code @ResourceLock} to a nested test class.
+	 */
+	default Set<Lock> provideForNestedClass(Class<?> testClass) {
+		return emptySet();
+	}
+
+	/**
+	 * Add shared resources to a test method.
+	 *
+	 * @param testClass a test class containing the {@code testMethod}
+	 * @param testMethod a test method to add shared resources
+	 * @return a set of {@link Lock}; may be empty
+	 *
+	 * @apiNote has the same semantics as adding {@code @ResourceLock} to a test method.
+	 */
+	default Set<Lock> provideForMethod(Class<?> testClass, Method testMethod) {
+		return emptySet();
+	}
+
+	/**
+	 *
+	 * <p>{@link Lock} represents a shared resource.
+	 *
+	 * <p>Each resource is identified by {@link #key}.
+	 * In addition,{@link #accessMode} allows you to specify
+	 * whether a test class or test
+	 * method requires {@link ResourceAccessMode#READ_WRITE READ_WRITE}
+	 * or only {@link ResourceAccessMode#READ READ} access to the resource.
+	 * In the former case, execution of the test will occur while no other test
+	 * that uses the shared resource is being executed.
+	 * In the latter case, the test class or test method
+	 * may be executed concurrently with other tests
+	 * that also require {@code READ} access but not at the same time
+	 * as any other test that requires {@code READ_WRITE} access.
+	 *
+	 * <p>This guarantee extends to lifecycle methods of a test class or method.
+	 * For example, if a {@code Lock} is added to a test method
+	 * then the "lock" will be acquired before any
+	 * {@link BeforeEach @BeforeEach} methods are executed and released after all
+	 * {@link AfterEach @AfterEach} methods have been executed.
+	 *
+	 * @since 5.12
+	 * @see Isolated
+	 * @see Resources
+	 * @see ResourceAccessMode
+	 * @see ResourceLock
+	 * @see ResourceLocksFrom
+	 * @see ResourceLocksProvider
+	 */
+	final class Lock {
```

Comment:
I wonder if it we should make this an interface and pass a `LockFactory` to the methods of `ResourceLocksProvider` rather than making the implementation part of the API. Having such a parameter object has proven useful in many cases but it might be overengineering here. Thoughts?

Reply:
Actually, I considered `Lock` like an immutable data carrier which has no behavior, that's why I'm not sure about an interface in this case.

Also, the current usage of `Lock` reminds `@ResourceLock`:
```java
@ResourceLock("key");
// both create a 'READ_WRITE' resource identified by 'key':
new Lock("key");
```

Maybe you have in mind some use cases for an interface?

Reply:
Not really, let's go with the record-like data class.